main.go

```diff
@@ -84,35 +87,98 @@ type Server struct {
+func (srv *Server) epochLoop(epochLen time.Duration) {
+	ticker := time.NewTicker(epochLen)
+	elog.Println("Starting epoch loop.")
+	for {
+		select {
+		case <-srv.done:
+			return
+		case <-ticker.C:
+			if err := srv.puncture(); err != nil {
+				if err.Error() == errEpochExhausted {
+					if err := srv.init(); err != nil {
+						elog.Fatal("Failed to re-initialize randomness server.")
+					}
+				}
+			}
+		}
+	}
 }
 
-// NewServer returns a new PPOPRF randomness server instance.
-//
-// FIXME Pass in a list of 8-bit tags defining epochs.
-// The instance will generate its own secret key.
-func NewServer() (*Server, *string, error) {
+// init (re-)initializes the randomness server instance of the Rust FFI.
+func (srv *Server) init() error {
+	srv.Lock()
+	defer srv.Unlock()
+
 	// FIXME should we runtime.LockOSThread() here?
 	raw := C.randomness_server_create()
 	if raw == nil {
-		return nil, nil, errors.New("failed to create randomness server")
+		return errors.New("failed to create randomness server")
 	}
-	server := &Server{raw: raw}
-	runtime.SetFinalizer(server, serverFinalizer)
+	srv.raw = raw
 
 	var pkOutput [serializedPkBufferSize]byte
 	pkSize := C.randomness_server_get_public_key(
-		server.raw, (*C.uint8_t)(unsafe.Pointer(&pkOutput[0])))
+		srv.raw, (*C.uint8_t)(unsafe.Pointer(&pkOutput[0])))
 	if pkSize == 0 {
-		return nil, nil, errors.New("failed to get public key")
+		return errors.New("failed to get public key")
 	}
-	pk := base64.StdEncoding.EncodeToString(pkOutput[:pkSize])
+	base64.StdEncoding.Encode(srv.pubKey, pkOutput[:pkSize])
+
+	elog.Println("(Re-)initialized server instance.")
+
+	return nil
+}
+
+// puncture punctures the randomness server's PPOPRF.  As part of the
+// puncturing, we're incrementing our epoch counter.  If we're about to exhaust
+// our counter (i.e., an integer overflow is about to happen), we return an
+// error, which signals to the caller that it's time to create a new randomness
+// server instance.
+func (srv *Server) puncture() error {
+	srv.Lock()
+	defer srv.Unlock()
+
+	C.randomness_server_puncture(srv.raw, (C.uint8_t)(srv.md))
+
+	// An epoch is exhausted when our 8-bit counter is about to overflow.
+	if srv.md == maxEpoch {
+		return errors.New(errEpochExhausted)
+	}
+	elog.Printf("Punctured epoch %d.", srv.md)
+	srv.md++
+	return nil
+}
+
+func serverFinalizer(server *Server) {
+	C.randomness_server_release(server.raw)
+	server.raw = nil
+}
+
+// NewServer returns a new PPOPRF randomness server instance.
+//
+// FIXME Pass in a list of 8-bit tags defining epochs.
+// The instance will generate its own secret key.
+func NewServer(epochLen time.Duration) (*Server, error) {
+	server := &Server{
+		done:   make(chan bool),
+		pubKey: make([]byte, base64.StdEncoding.EncodedLen(int(serializedPkBufferSize))),
+	}
+	if err := server.init(); err != nil {
+		return nil, err
+	}
+	runtime.SetFinalizer(server, serverFinalizer)
+	go server.epochLoop(epochLen)
```

Comment:
Another method:

On each randomness request, we could execute `getEpoch` and compare the result to `srv.md`. If they are different, then puncturing process can be triggered and `srv.md` can be updated.

Reply:
Right, I forgot about that. Commit 591df21 should address this issue.

Reply:
> On each randomness request, we could execute getEpoch and compare the result to srv.md. If they are different, then puncturing process can be triggered and srv.md can be updated.

I like this idea because it's simpler to implement. I'll give this some more thought, and may address it in a follow-up PR.